### PR TITLE
Issue #90 Policy based access control for OIDC OP

### PR DESCRIPTION
**Pull request by Iwakata**
## Analysis

This feature provides policy base access control on OIDC and OAuth 2.0 endpoints.  It is a counterpart of the issue #89 "Policy based access control for SAML 2.0 IdP ".

## Solution

The main part of this feature is `checkPolicy` method defined in `ResourceOwnerSessionValidator.java`.  It checks session/user properties of the user, in this case `ResourceOwner`,  against the applicable access policy.  Some grant types that do not have user's session are denied for access.  Those grant types are:

- Password
- PasswordCredentials
- SAML2Bearer

This feature can be turned on or off for each OAuth client.

## Install/Update

1. **Enabling the feature**

In order to enable this feature, go to Agent -> OAuth2 OIDC Client page and add a check mark to "Policy Based Endpoint Protection".

2. **Defining a new resource type**

Click  "New Resource Type" button and specify the following:
- Name:  any name of the resource type
- Patterns: "client_id=*"
- Actions: "IssueToken" 

And the click on the "Save changes" button.

3. **Defining a new policy set**

Click "New Policy Set" button and specified the following:

- Id: "OAuthClientService"
- Name: any name of the policy set
- Resource Types:  the resouce type name specified above

4. **Adding policies to the policy set**

Add a policy or policies with regard to the resource type specified above.

## Compatibility

None

## Performance

Performance might be affected if complicated access policies need to be applied.

## I18N

None

## Testing

- Set up an OP of OAuth2.0/OIDC.
- Enable this feature and define the resource type and policies.
- Request a token.

Depending on the user/session properties, the request is:

- denied or
- accepted and a token is issued or
- processed further and session upgrade is performed.

## Regression testing

None

### CODE REVIEW

**Comment by Iwakata** [comment]
Currently "Membership Condition" does not work.  This pull request is partially dependent on the pull request Number 170.  If both 170 and 173 are merged, then "Membership condition" will work.

File: openam-oauth2/src/main/java/org/forgerock/oauth2/core/ClientRegistration.java [hunk 1]
```diff
@@ -149,4 +150,12 @@
      * @return true if the client is configured to skip resource owner consent.
      */
     boolean isConsentImplied();
+
+     /**
+      * Gets whether or not Policy Based Endpoint Protection is enabled.
+      *
+      * @return true if enabled.
+      */
+     boolean isPolicyBasedProtectionEnabled();
```
Comment (tsujiguchitky) [inline openam-oauth2/src/main/java/org/forgerock/oauth2/core/ClientRegistration.java]: Please check the indent.

File: openam-oauth2/src/main/java/org/forgerock/oauth2/core/ResourceOwnerSessionValidator.java [hunk 2]
```diff
@@ -371,6 +394,35 @@ private ResourceOwnerAuthenticationRequired authenticationRequired(OAuth2Request
...
         return new ResourceOwnerAuthenticationRequired(loginUrl);
     }
+    
+    private ResourceOwnerAuthenticationRequired additionalAuthenticationRequired(OAuth2Request request, Parameter param)
+            throws AccessDeniedException, URISyntaxException, ServerException, NotFoundException,
+            UnsupportedEncodingException {
+        OAuth2ProviderSettings providerSettings = providerSettingsFactory.get(request);
+        Template loginUrlTemplate = providerSettings.getCustomLoginUrlTemplate();
+
+        removeLoginPrompt(request.<Request>getRequest());
```
[the remark this answers is not in the document]
Comment (tsujiguchitky) [reply openam-oauth2/src/main/java/org/forgerock/oauth2/core/ResourceOwnerSessionValidator.java]: The next line has same issue.

File: openam-oauth2-saml2/src/main/java/org/forgerock/openam/oauth2/saml2/core/Saml2GrantTypeHandler.java [hunk 3]
```diff
@@ -12,10 +12,12 @@
  * information: "Portions copyright [year] [name of copyright owner]".
  *
  * Copyright 2014-2016 ForgeRock AS.
+ * Portions Copyrighted 2019 Open Source Solution Technology Corporation.
  */
 
 package org.forgerock.openam.oauth2.saml2.core;
 
+import org.forgerock.oauth2.core.SAML2BearerRequestValidator;
```
[the remark this answers is not in the document]
Comment (Iwakata) [reply openam-oauth2-saml2/src/main/java/org/forgerock/openam/oauth2/saml2/core/Saml2GrantTypeHandler.java]: I will fix this.

File: openam-oauth2/src/main/java/org/forgerock/oauth2/core/ResourceOwnerSessionValidator.java [hunk 4]
```diff
@@ -515,6 +567,90 @@ private String getAuthURL(HttpServletRequest request) {
                 + "/UI/Login";
     }
 
+        /**
```
[the remark this answers is not in the document]
Comment (Iwakata) [reply openam-oauth2/src/main/java/org/forgerock/oauth2/core/ResourceOwnerSessionValidator.java]: I will fix this.

File: openam-oauth2/src/main/java/org/forgerock/oauth2/core/ClientRegistration.java (hunk 1 above)
Comment (Iwakata) [reply openam-oauth2/src/main/java/org/forgerock/oauth2/core/ClientRegistration.java]: I will fix this.

File: openam-oauth2/src/main/java/org/forgerock/oauth2/core/ResourceOwnerSessionValidator.java [hunk 5]
```diff
@@ -92,6 +92,15 @@
 import org.restlet.data.Reference;
 import org.restlet.ext.servlet.ServletUtils;
 
+// Added for checkPolicy()
+import com.sun.identity.entitlement.Entitlement;
+import com.sun.identity.entitlement.EntitlementException;
+import com.sun.identity.entitlement.Evaluator;
+import com.sun.identity.entitlement.opensso.SubjectUtils;
+import com.sun.identity.idm.IdRepoException;
+import com.sun.identity.policy.interfaces.Condition;
```
[the remark this answers is not in the document]
Comment (Iwakata) [reply openam-oauth2/src/main/java/org/forgerock/oauth2/core/ResourceOwnerSessionValidator.java]: I agree.

File: openam-oauth2/src/main/java/org/forgerock/oauth2/core/ResourceOwnerSessionValidator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Iwakata) [reply openam-oauth2/src/main/java/org/forgerock/oauth2/core/ResourceOwnerSessionValidator.java]: Understood.

**Pull request merged**